numpy/core/src/multiarray/datetime.c

```diff
@@ -1761,6 +1762,9 @@ parse_datetime_unit_from_string(char *str, Py_ssize_t len, char *metastr)
                 return NPY_FR_as;
         }
     }
+    else if (allow_generic && !strcmp(str, "generic")) {
```

Comment:
I'm surprised the string is not necessarily zero-terminated, though I'll take your word for it :-)